central/compliance/datastore/internal/store/postgres/store.go

```diff
@@ -246,14 +270,63 @@ func (s *storeImpl) StoreComplianceDomain(ctx context.Context, domain *storage.C
 	return nil
 }
 
-func (s *storeImpl) StoreAggregationResult(_ context.Context, _ string, _ []storage.ComplianceAggregation_Scope, _ storage.ComplianceAggregation_Scope, _ []*storage.ComplianceAggregation_Result, _ []*storage.ComplianceAggregation_Source, _ map[*storage.ComplianceAggregation_Result]*storage.ComplianceDomain) error {
+func (s *storeImpl) StoreAggregationResult(ctx context.Context, query string, groupBy []storage.ComplianceAggregation_Scope, unit storage.ComplianceAggregation_Scope, results []*storage.ComplianceAggregation_Result, sources []*storage.ComplianceAggregation_Source, domainMap map[*storage.ComplianceAggregation_Result]*storage.ComplianceDomain) error {
+	scopeTree, err := effectiveAccessScopeTree(ctx)
+	if err != nil {
+		return err
+	}
+	key := aggKey(groupBy, unit, query, scopeTree)
+	value := aggCacheEntry{
+		results:   results,
+		sources:   sources,
+		domainMap: domainMap,
+	}
+	s.aggregatedDataCache.Add(key, value)
 	return nil
 }
 
-func (s *storeImpl) GetAggregationResult(_ context.Context, _ string, _ []storage.ComplianceAggregation_Scope, _ storage.ComplianceAggregation_Scope) ([]*storage.ComplianceAggregation_Result, []*storage.ComplianceAggregation_Source, map[*storage.ComplianceAggregation_Result]*storage.ComplianceDomain, error) {
-	return nil, nil, nil, nil
+func (s *storeImpl) GetAggregationResult(ctx context.Context, query string, groupBy []storage.ComplianceAggregation_Scope, unit storage.ComplianceAggregation_Scope) ([]*storage.ComplianceAggregation_Result, []*storage.ComplianceAggregation_Source, map[*storage.ComplianceAggregation_Result]*storage.ComplianceDomain, error) {
+	scopeTree, err := effectiveAccessScopeTree(ctx)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	key := aggKey(groupBy, unit, query, scopeTree)
+	value, ok := s.aggregatedDataCache.Get(key)
+	if !ok {
+		return nil, nil, nil, nil
+	}
+	return value.results, value.sources, value.domainMap, nil
 }
 
 func (s *storeImpl) ClearAggregationResults(_ context.Context) error {
+	s.aggregatedDataCache.Purge()
 	return nil
 }
+
+func effectiveAccessScopeTree(ctx context.Context) (*effectiveaccessscope.ScopeTree, error) {
+	targetResource := resources.Compliance
+	scopeChecker := sac.GlobalAccessScopeChecker(ctx).AccessMode(storage.Access_READ_ACCESS).Resource(targetResource)
+	scopeTree, err := scopeChecker.EffectiveAccessScope(permissions.View(targetResource))
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get effective access scope")
+	}
+	return scopeTree, nil
+}
+
+func aggKey(groupBy []storage.ComplianceAggregation_Scope, unit storage.ComplianceAggregation_Scope, query string, scopeTree *effectiveaccessscope.ScopeTree) aggCacheKey {
+	sort.SliceStable(groupBy, func(i, j int) bool {
+		return groupBy[i] < groupBy[j]
+	})
+	key := fnv.New64a()
+	addKey(key, unit.String())
+	addKey(key, query)
+	addKey(key, scopeTree.Compactify().String())
```

Comment:
The `String()` function on `ScopeTreeCompacted` does a map iteration. The keys there are not guaranteed to come up in a deterministic order.
It would be good to ensure the aggregated key is the same for the same scope tree.